File: 2020/day-01/part2.rs
fn main() {
    let contents: String = std::fs::read_to_string("input.txt")
            .expect("failed to read file");
    let values: Vec<i32> = contents
            .split("\n")
            .map(|s| s.parse::<i32>().unwrap_or(0))
            .filter(|n| n.is_positive())
            .collect();
    for i in 0 .. values.len() {
      for j in i+1 .. values.len() {
        for k in j+1 .. values.len() {
          if values[i] + values[j] + values[k] == 2020 {
            print!("{}\n", values[i] * values[j] * values[k]);
          }
        }
      }
    }
}

File: 2021/day-04/part2.hs
import Data.List
import Data.Maybe

reshape n l = if length l <= n then [l] else (take n l) : (reshape n $ drop n l)

wordsBy p l = case findIndex p l of
    Nothing -> [l]
    Just n -> let (h, t) = splitAt n l in (h:(wordsBy p $ tail t))
readInt = read :: String -> Int
parseBoards = (map (map (map readInt . words) . tail)) . reshape 6
parseNumbers = map readInt . (wordsBy (== ','))

setNumber n x = if n == x then -1 else x
markNumber = map . map . setNumber

checkRow = (== -5) . foldl1 (+)
checkRows = foldl1 (||) . map checkRow
checkBoard b = checkRows b || (checkRows $ transpose b)

score :: [[Int]] -> Int
score = (foldl1 (+)) . (map (foldl1 (+))) . ((map . map . max) 0)

step :: Int -> [[[Int]]] -> [[[Int]]]
step = map . markNumber
stepAndFilter num = (filter (not . checkBoard)) . (step num)
lastWinningScore nums boards = 
    let bs = stepAndFilter (head nums) boards in if length bs == 0
        then head nums * (score $ markNumber (head nums) $ fromJust $ find (not . checkBoard) boards)
        else lastWinningScore (tail nums) bs


parse s = let ls = lines s in (parseNumbers $ head ls, parseBoards $ tail ls)

main = do
  contents <- readFile "input.txt"
  print $ uncurry lastWinningScore $ parse contents
